### DISCUSSION TITLE
A path other than / and retaining the scope

### DISCUSSION THREAD

**Discussion by gabrielschulhof** [verdaccio/verdaccio#4940] [Q&A] [answered]
Can we add a configuration option to verdaccio so that it serves packages from a path other than /, and so it retains the scopes of packages that are published to it when serving them back? IOW, I'd like to be able to set my registry as http://localhost:4873/some/path instead of http://localhost:4873/, and I would like the name of the tarball of packages served by verdaccio to include the scope.

Motivation: I was able to set up my laptop with verdaccio and nginx to make it look like our company's jfrog repo:

1. Set up verdaccio with the jfrog repo as the uplink.
2. In /etc/hosts, set the JFrog domain as 127.0.0.1.
3. Run nginx locally, and proxy the jfrog domain to verdaccio, including slicing the path `/artifactory/api/npm/npm-repo` off the incoming path.
4. Run verdaccio with a Node.js preload (`node -r verdaccio.preload.js $(which verdaccio) -c /path/to/config/file.yml`) that overrides dns.lookup, giving the real IP for the JFrog domain instead of 127.0.0.1.

With these steps, my configuration can point to what looks like our company jfrog domain, while in reality it goes through verdaccio running locally. Indeed, installing dependencies with `yarn` runs through verdaccio and works just fine.

The problem is, when I run `yarn up <package_name>`, the resulting yarn.lock has the wrong paths:

```yaml
-  resolution: "@company/package@npm:3.1.200::__archiveUrl=https%3A%2F%company.jfrog.io%2Fartifactory%2Fapi%2Fnpm%2Fnpm-repo%2F%40company%2Fpackage%2F-%2F%40company%2Fpackage-3.1.200.tgz"
+  resolution: "@company/package@npm:3.1.206::__archiveUrl=https%3A%2F%company.jfrog.io%2F^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^%40company%2Fpackage%2F-%2F^^^^^^^^^^^^^package-3.1.206.tgz"
```

where the `^^^` sequences are missing.

With the configuration options I request, I could upgrade packages such that the resulting yarn.lock changes could be committed, because they continue to point at our private registry.

**Comment by gabrielschulhof** [comment]
OK, FWIW, I set

```yaml
url_prefix: /artifactory/api/npm/npm-repo/
```

in the config file, and I ran verdaccio with the environment variable

```bash
VERDACCIO_PUBLIC_URL='https://company.jfrog.io/artifactory/api/npm/npm-repo
```

and when I ran `yarn up <package_name>`, the result was that it completely removed the `::__archiveUrl=...` portion from the yarn.lock file.

I tested the result with direct access to our actual company jfrog restored, and it seemed able to update. So, this is kind of a workaround, however, I'd still rather emulate the way yarn 4 interacts with our company jfrog registry.

**Comment by gabrielschulhof** [comment]
Setting things up as described in my [previous comment](https://github.com/verdaccio/verdaccio/issues/4936#issuecomment-2461480437) seems to break the UI. If I browse to https://localhost:4873/ I get a blank screen, where before I would get a list of packages that are published, or a message saying that no packages are published.

**Comment by gabrielschulhof** [comment]
OK, it's because verdaccio is running with a self-signed certificate. I can make a security exception for it and it's all good.

**Comment by gabrielschulhof** [comment]
With the `url_prefix` and the `VERDACCIO_PUBLIC_URL` set as described above, the page does not display any packages that I publish

<img width="1155" alt="image" src="https://github.com/user-attachments/assets/90ec695e-1dbd-43ab-93d6-558640aa13ce">

even though the publishing of the packages succeeded.

**Comment by gabrielschulhof** [comment, marked as answer]
OK, looks like I had to type the proxy URL into the location, not the actual URL (i.e. https://company.jfrog.io instead of https://localhost:4873/) in order to get the list of published packages.

**Reply by mbtools** [reply]
You got it. VERDACCIO_PUBLIC_URL specifies the domain you need for your browser. 

Anything else? if not, maybe mark this as answered (by yourself 😉 )
